### ISSUE TITLE
[$2000] Split bill shows alphabets for very brief moment when pasted the copied text

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#18562]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Copy some alphabetical text (i.e.: Abcd)
2. Go to split bill
3. Paste the copied text in split bill amount

## Expected Result: 
Split should restrict alphabets completely
## Actual Result: 
Split bill shows alphabets for a bit when pasted.

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.11-2 
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/236699096-3f36ac00-ddcc-4a94-b1b1-bca06900c17e.MP4

https://user-images.githubusercontent.com/43996225/236699104-3a2867c4-b15f-4e47-8f70-92e3294a07f9.MP4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @DinalJivani
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1683416584966559

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01569ec2d58fbfa63f</li>
        <li>Upwork Job ID: 1655626738889830400</li>
        <li>Last Price Increase: 2023-05-30</li>
    </ul>
</details>

**Comment by arielgreen** [comment]
Reproduced following steps in OP

![2023-05-08_13-29-23 (1)](https://user-images.githubusercontent.com/37308300/236891044-4b911249-e731-465f-9fc4-73f1a41d0bf1.gif)

**Comment by arielgreen** [comment]
@parasharrajat  @youssef-lr can you confirm?

**Comment by arielgreen** [comment]
Whether this is an Expensify bug or a larger [issue with RN](https://github.com/Expensify/App/issues/18562#issuecomment-1539590848).

**Comment by arielgreen** [comment]
Going to bump the price here.

**Comment by akinwale** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The text input field in the Split bill / Send money / Request money screens briefly shows alphabets when copied text is pasted into the field.

### What is the root cause of that problem?
There is no check for pasted text in the underlying native UITextField on iOS, thereby resulting in the pasted text to display before the value is actually set by the React Native implementation. 

### What changes do you think we should make in order to solve the problem?
The underlying native UITextField implementation has to be updated to prevent pasting arbitrary text based on the keyboardType. This can be achieved with a few lines of Objective-C using the UITextField delegate method [shouldChangeCharactersInRange](https://developer.apple.com/documentation/uikit/uitextfielddelegate/1619599-textfield).

This can be fixed by adding the following lines to the delegate implementation in https://github.com/facebook/react-native/blob/2b932c3820cfabef1590f6fcf8b95f65d9c21eb8/packages/react-native/Libraries/Text/TextInput/RCTBackedTextInputDelegateAdapter.m#L75-L77. The lines can be added after the check for `newText` being `nil`.

```
UIKeyboardType keyboardType = textField.keyboardType;
if (keyboardType == UIKeyboardTypeNumberPad || keyboardType == UIKeyboardTypeDecimalPad) {
  // Check if the new text is in the numeric character set when the keyboard type is numeric / decimal.
  // With this check, updates to the text field can be prevented in cases where arbitrary text is pasted.
  NSCharacterSet *numericCharSet = [NSCharacterSet decimalDigitCharacterSet];
  if (![[newText stringByTrimmingCharactersInSet:numericCharSet] isEqualToString:@""]) {
    return NO;
  }
}
```

These changes can be made to Expensify's react native fork. It can also be proposed as a PR to upstream React Native if desired.

### What alternative solutions did you explore? (Optional)
None.

**Comment by Thanos30** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

React Native TextInput flickering on formatting the input.

### What is the root cause of that problem?

The root cause of this problem is the nature of React Native, which is also the reason why this doesn't happen on the web version.

The TextInput component has a value of the formatted text (the user's input formatted to accept only numbers and points), but it also works as a basic input field that accept the original input. These split seconds in between the formatting and re-rendering of the component has as a result that annoying flicker.
`<TextInput
            disableKeyboard
            autoGrow
            hideFocusedState
            inputStyle={[styles.iouAmountTextInput, styles.p0, styles.noLeftBorderRadius, styles.noRightBorderRadius]}
            textInputContainerStyles={[styles.borderNone, styles.noLeftBorderRadius, styles.noRightBorderRadius]}
            onChangeText={props.onChangeAmount}
            ref={props.forwardedRef}
            value={props.formattedAmount}    <------ value mentioned above.
            placeholder={props.placeholder}
            keyboardType={CONST.KEYBOARD_TYPE.NUMBER_PAD}
            blurOnSubmit={false}
            selection={props.selection}
            onSelectionChange={props.onSelectionChange}
        />`

### What changes do you think we should make in order to solve the problem?

The way I counter this problem in many similar cases, is by using 2 different TextInput fields. One will work as the display, and will not accept user input, ad the other one will accept user input, but it won't be shown.

So basically, we will have a TextInput that will only get the value of the formatted text, without the original text, and a TextInput that will receive the user Input, but it won't be visible. The transparent TextInput will have an absolute position and will be put exactly where the display will be, so this will look smoothly.

The AmountTextInput component we have, will now be consisted of 2 TextInputs the way I mentioned above.

Video of solution:

https://www.loom.com/share/abd05c603fb94d0f95af7a109daf5322

### What alternative solutions did you explore? (Optional)
Played around for some time, haven't figured out any other way that doesn't require to edit files outside of our repo.

**Comment by arielgreen** [comment]
@parasharrajat when can we expect these proposals to be reviewed?

**Comment by youssef-lr** [comment]
bump @parasharrajat! :)

**Comment by akinwale** [comment]
> @akinwale I need more explanation to understand your proposal. IMO both the root cause and the solution is not clearly explained.
> 
> To me, it seems like you are suggesting controlling the formatting on the native layer.
> 
> Lacking elements:
> 
> 1. Why are you proposing these changes?

I am proposing these changes because I've explored other options (which aren't workarounds) within React Native which did not solve the flickering problem. Some examples include:
- Using `event.nativeEvent.preventDefault()` with the `onChange` event handler
- Implementing an `onChangeText` to prevent any modifications to the input
- Making sure the state variable set in the value is updated in the same component.

With iOS native, there are delegate methods designated for editing the text in a `UITextInput` control, which are provided by the `UITextFieldDelegate` protocol. One of these delegate methods is the `textField(_:shouldChangeCharactersIn:replacementString:)`. Since the delegate method is already implemented in React Native, all that's needed is to update it to handle this scenario.

This is a quote from the discussion section from the [Apple developer documentation](https://developer.apple.com/documentation/uikit/uitextfielddelegate/1619599-textfield#discussion) for this method indicating what it's suitable for.
> The text field calls this method whenever user actions cause its text to change. Use this method to validate text as it is typed by the user. For example, you could use this method to prevent the user from entering anything but numerical values.

The stated example fits our use case perfectly.

> 2. How will this work overall? Impact of changes on TextInput behavior?

As indicated in the documentation, the method implementation, will only execute whenever a user makes changes to the text input control (either by typing or pasting into the control).

With the proposed changes, it will only execute on text fields which have one of the following keyboard types set: `UIKeyboardTypeDecimalPad` or `UIKeyboardTypeNumberPad`. More keyboard types can be added to the conditional check if required.

The `TextInput` will still behave as expected for other keyboard types, since the implementation remains the same. Delegates in iOS are simply used to modify behaviour of an object or control in certain scenarios / conditions, but not the core functionality of that object.


> 3. Why hasn't this been done on RN yet? Any related issues research material..

I cannot speak to this. Perhaps, it's never been a priority, and I have been unable to find anything related to this yet. I will update this post if I make any new discoveries.

**Comment by akinwale** [comment]
> @akinwale What about Android?

Wow. I didn't realise this was also happening on Android. I'm actually a bit surprised React Native doesn't handle this out of the box.

We'd have to make a similar change to the Android native control (in Java code) in one or both of these files:
* [ReactEditText.java](https://github.com/facebook/react-native/blob/671ea383fe45dd9834a0c0481360de050df7f0c9/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java)
* [ReactTextInputManager.java](https://github.com/facebook/react-native/blob/671ea383fe45dd9834a0c0481360de050df7f0c9/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java)

Interestingly, after going through the source, it looks like an `InternalKeyListener` was implemented which permits all keyboard input regardless of the type of the keyboard that is displayed, which explains why any text can be pasted at this point.

https://github.com/facebook/react-native/blob/671ea383fe45dd9834a0c0481360de050df7f0c9/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java#L1222

Usually on Android, an EditText with input type set to numeric or decimal would only allow pasting numbers.

I'll need to investigate some more to figure out what modifications would work in this case.

**Comment by akinwale** [comment]
> @akinwale What happens natively if a decimal type of keyboard is used on Android and iOS? Do they show this flicker when pasting the text when only the keyboard type is used with no conditional logic to prevent characters?

@parasharrajat These are based on my tests with new projects in both Android Studio and XCode.

On Android, with the `inputType` for an `EditText` set to `number` or `numberDecimal`, nothing happens when there is an attempt to paste a string containing alphabets. As noted above, React Native overrides this behaviour to enable any input.

On iOS, with the `keyboardType` for a `UITextField` set to `Number Pad` or `Decimal Pad`, any pasted text (including alphabets) is accepted by the text field.

**Comment by arielgreen** [comment]
Reassigning, OOO tomorrow and Monday.

**Comment by alexpensify** [comment]
@youssef-lr - I would love your thoughts here. @parasharrajat has some good points but I want to get a confirmation before closing. Thanks!

**Comment by alexpensify** [comment]
No update

**Comment by alexpensify** [comment]
@arielgreen - I think you are back online. Let me know when I can unassign myself. Thanks!

**Comment by alexpensify** [comment]
@youssef-lr - we still need feedback, here is my post from last week: https://github.com/Expensify/App/issues/18562#issuecomment-1565137369

**Comment by youssef-lr** [comment]
Sorry for missing your comment @alexpensify. Yeah I agree we can close this out as we discussed in Slack.

**Comment by alexpensify** [comment]
All good, thanks!
